### ISSUE TITLE
Silent Push Notifications/Receiving Notifications in Background

### ISSUE THREAD

**Issue by kevindonahue** [phonegap-build/PushPlugin#288]
We have managed to get this plugin to respond to silent push notifications which in turn call javascript functions, all while running in background mode.  I thought we would share our solution as it seems like a lot of people are wanting this feature.

In AppDelegate+notification.m, change didReceiveRemoteNotification:

```
- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo fetchCompletionHandler:(void (^)(UIBackgroundFetchResult result))handler
{
    NSLog(@"didReceiveNotification");

    // Get application state for iOS4.x+ devices, otherwise assume active
    UIApplicationState appState = UIApplicationStateActive;
    if ([application respondsToSelector:@selector(applicationState)]) {
        appState = application.applicationState;
    }

    PushPlugin *pushHandler = [self getCommandInstance:@"PushPlugin"];
    pushHandler.notificationMessage = userInfo;
    if (appState == UIApplicationStateActive)
        pushHandler.isInline = YES;
    else
        pushHandler.isInline = NO;
    [pushHandler notificationReceived];

    handler(UIBackgroundFetchResultNewData);
}
```

In Pushplugin.m, change notificationReceived:

```
- (void)notificationReceived {
    NSLog(@"Notification received");

    if (notificationMessage && self.callback)
    {
        NSMutableString *jsonStr = [NSMutableString stringWithString:@"{"];

        [self parseDictionary:notificationMessage intoJSON:jsonStr];

        if (isInline)
        {
            [jsonStr appendFormat:@"foreground:\"%d\"", 1];
            isInline = NO;
        }
        else
            [jsonStr appendFormat:@"foreground:\"%d\"", 0];

        [jsonStr appendString:@"}"];

        NSLog(@"Msg: %@", jsonStr);

        NSString * jsCallBack = [NSString stringWithFormat:@"%@(%@);", self.callback, jsonStr];
        [self.webView stringByEvaluatingJavaScriptFromString:jsCallBack];

        //get javascript function to fire in background mode
        CDVPluginResult *commandResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:jsonStr];
        [self.commandDelegate sendPluginResult:commandResult callbackId:self.callbackId];

        self.notificationMessage = nil;
    }
}
```

Cheers!

**Comment by cburghardt** [comment]
I have tested the proposed changes and also adopted the plist but the Javascript is still not called when the app is in background (IOS 8). It only works in foreground.
One comment: handler(UIBackgroundFetchResultNewData); should be called when all the work has been done so you probably need a callback when the Javascript has finished - similar to the background fetch implemented here: https://github.com/christocracy/cordova-plugin-background-fetch

**Comment by JeremyColton** [comment]
I am using this on ios 7.1. I send a push notification to my phone and in the XCode logs I see the app receives the notification, the JS callback is fired, but it exits part way through. It doesn't run any lengthy code, just writes to the log to say it's doing something.  When I click on the app's icon, the JS resumes where it stopped!  Has anyone else seen this behaviour?

**Comment by huttj** [comment]
What about for Android? Any work on that?

**Comment by JeremyColton** [comment]
The ecb callback is firing, but it stops after running a few lines of javascript as if the webview thread was given a few microseconds to run and is then paused.  When I then click on my app's icon to bring it back into the foreground, the JS method continues where it left off!

**Comment by kevindonahue** [comment]
Hi Jeremy,

Its been a long time since I posted this original code and we've learned a lot since then.  What you are experiencing is caused by this line:

handler(UIBackgroundFetchResultNewData);

It is actually a timing issue.  By calling this line you are telling iOS that you are finished with the execution of the background task.  If you have a lengthy javascript process, it can interrupt it with this statement, because iOS may have already put the App back to sleep after this call.

What we did to prevent this is to create a singleton class in the plugin and make the return handler a property of the singleton:

Create a singleton class, ours was called VBSingleton.h:

```
#import <Foundation/Foundation.h>

@interface VBSingleton : NSObject

@property (strong,nonatomic) void (^backgroundHandler)(UIBackgroundFetchResult result);

+ (VBSingleton*)sharedInstance;

@end
```

Also need a VBSingleton.m class:

```
#import "VBSingleton.h"

@implementation VBSingleton

+ (VBSingleton*)sharedInstance
{
    static dispatch_once_t pred = 0;
    __strong static id _sharedObject = nil;
    dispatch_once(&pred, ^{
        NSLog(@"%@",@"VBSingleton Shared first call");
        _sharedObject = [[VBSingleton alloc] init];
    });
    return _sharedObject;
}

@end
```

Then you no longer call this line at the end of didReceiveRemoteNotification: 
```
handler(UIBackgroundFetchResultNewData);
```
and instead call this line:
```
[VBSingleton sharedInstance].backgroundHandler = handler;
```

So, we're basically saving the handler in a global variable to be executed at a later time.  So, once your javascript has finished running, you then call a plugin method which calls the background complete handler:

Add a new method in the Plugin.h class:
```
- (void)didCompleteBackgroundProcess:(CDVInvokedUrlCommand*)command;
```
Declared as follows in Plugin.m:
```
- (void)didCompleteBackgroundProcess:(CDVInvokedUrlCommand*)command {
    NSLog(@"didCompleteBackgroundProcess called - calling background handler %lx",(unsigned long)[VBSingleton sharedInstance].backgroundHandler);
    if ([VBSingleton sharedInstance].backgroundHandler != nil) {
        [VBSingleton sharedInstance].backgroundHandler(UIBackgroundFetchResultNewData);
        [VBSingleton sharedInstance].backgroundHandler = nil;
    }
}
```
Add a new javascript function to PushNotification.js:

```
PushNotification.prototype.backgroundDone = function(successCallback, errorCallback) {
    if (errorCallback == null) { errorCallback = function() {}}

    if (typeof successCallback != "function") {
        console.log("PushNotification.backgroundDone failure: success callback parameter must be a function");
        return
    }

     cordova.exec(successCallback, errorCallback, "PushPlugin", "didCompleteBackgroundProcess", []);
};
```

Just call the backgroundDone javascript method when you have finished processing.

Just be aware that iOS only give you a finite time to run processes in the background.  The longer you take, the less likely iOS is to wake you up in the future.

From our experience with Push Notifications, they are not a very reliable communication network.  iOS can refuse to call the didReceiveRemoteNotification method if:
* it thinks you are draining too much battery
* your cellular network reception is poor
* you are abusing the background process execution time or are sending push messages too frequently
* other unexplainable reasons

You will notice this behaviour particularly when your phone is not plugged in/versus plugged in.

Hopefully Apple improve this technology soon.

Hope this helps :)

**Comment by JeremyColton** [comment]
Thanks so much Kevin, that makes sense and works!

How many messages is too many in your experience?  I will be sending around 10-100 per user over a few hours...

**Comment by kevindonahue** [comment]
Honestly, I wish I could tell you.  There doesn't seem to be any consistency in it.  We had numerous iPhones getting the same amount of push notification in the same period.  Some of them stopped working before others.  

I sent numerous messages to Apple trying to get insights on how this process works.  The response from Apple was that the formula behind it was a secret not meant for public knowledge.

Basically, what we worked out is there are numerous factors which can affect whether or not iOS decides to call the didReceiveRemoteNotification:

* Phone plugged in versus running on battery
* Phone running on wifi versus running on cellular data
* The strength of your cellular data connection
* Amount of time on average your app is woken up from background
* Number of messages sent per hour
* The model of the iPhone used and iOS version
* And possibly some other variables we haven't worked out yet

Here are some discussions we've been having regarding issues we found with silent push notifications:

https://devforums.apple.com/message/1090707#1090707
http://stackoverflow.com/questions/26959472/silent-push-notifications-only-delivered-if-device-is-charging-and-or-app-is-for

Bottom line: silent push notifications are not currently a reliable communication foundation

**Comment by JeremyColton** [comment]
Thanks for all this wealth of info!  iOS is very irritating especially since Android's GCM notifcation system is so reliable for real-time updates!

So for iOS you use the following strategy to update the phone with realtime data:
1. your app polls the server every X seconds for updated data.
2. Use a push notification just to signal to the client to poll the server 'now' instead of waiting up to X seconds.

Correct?

**Comment by kevindonahue** [comment]
We wondered the same about WhatsApp, but the difference is that the push messages aren't silent.  We noticed that when looking at our iOS logs, pretty much all our silent notification actually get received by the iPhone, its just that iOS chooses not to wake up the App based on the variables I specified before.  

Android's GCM is very reliable, but even it has some throttling.  We found that if we alternated our data messages amongst 4 different buckets, we got around its throttling mechanisms.

Polling via background fetch seems even less reliable.  Sure you can specify a period of how often you would "like" to be woken up, but again iOS has some sort of "smart" formula built in where it decides how often it thinks you "need" to be woken up.  So even if you say you would like to perform a fetch request every minute, you will find that iOS slowly changes this over time to the point that it might wake you up every hour or two (or worse).

So, whats the solution for a reliable real time data app that fits within Apple's rules?  Good question! :)

**Comment by JeremyColton** [comment]
So data cannot be embedded in the notification.  Instead, when iOS wakes up the app after receiving a silent notification, the app can poll the server instantly so it's more up-to-date when the user next puts the app in the foreground.  This will save the app from needing to update itself for a given time period when the app returns to the foreground.

The question becomes what is the best polling method - every X seconds via javascript or native or use a Web Socket?

**Comment by kevindonahue** [comment]
Yes, that should be fine as long as you don't need the data until the app is brought to the foreground.  If your app is sending the real time data to a peripheral device, then you will be in trouble using push notifications because of the unreliability of iOS waking up your app from the background and calling the didReceiveRemoteNotification method, which is what our situation was :)

I would recommend using websockets, as you only need to transfer data when an update occurs on the server. I recommend looking at Pusher, as it has a great javascript library for websockets.

**Comment by JeremyColton** [comment]
i don't use a peripheral device. My app needs real-time data at certain times of its usage.  Either way, since users can opt out of push notifications, we must build a backup mechanism so the websockets route seems the best solution.  Thanks so much for your time and effort.

**Comment by drakula84** [comment]
I understand this patch works when the application is in background mode. But was anybody able get it working when application is in suspended mode? Or is supposed to work in suspended mode?

I see that in notificationReceived() self.callback is null so it is the end of thread.. Does anybody have an idea on how to start the Cordova application upon a silent push notifications so that you can deliver the notification to the JavaScript layer?

**Comment by tuomaslaa** [comment]
@kevindonahue can you please post/upload whole modified PushPlugin source code.

**Comment by jbeuckm** [comment]
Here it is with the changes. Should I do a pull request? https://github.com/jbeuckm/PushPlugin

**Comment by tuomaslaa** [comment]
@jbeuckm @kevindonahue  Got it working. Thanks. But I ran to a new problem. When sending a silent notification and the app is not running (iOS has ended the process to free memory resources) it will restart the app and it'll make push notification registeration at the same time with the silent notification processing. Now the push notification "register" function won't return the correct deviceToken. Instead it returns the content of the silent notification like '{"content-available":"1","data":"lorem lipsum","type":"event_type",foreground:"0"}'. How can I prevent this?

**Comment by ronaldjeremy** [comment]
@tuomaslaa thank you for mentioning this, I had this happen in my app and it really scared me because i couldn't pinpoint its cause.  My app registered that json string as its deviceToken to the server and effectively cut itself off from future notifications.  So i added some filtering server side to ignore a deviceToken with brackets in the string...

**Comment by ToniKorin** [comment]
Hi all users! Great work! How are you experienced this new functionality so far? Any obstacles to just publish this branch at http://plugins.cordova.io/? Btw, are those changes compatible also with the old non silent notifications?

**Comment by tuomaslaa** [comment]
It seems that the handler won't get fired with the iOS 6 even in the foreground scenario. Has anyone noticed same behavior? There is only one percent using iOS 6 anymore but still it could be great if it would work.

**Comment by ToniKorin** [comment]
So far my experience about reliability is pretty much following @kevindonahue's comments. The strength of your cellular data connection have to be really good that your app gets some execution time in the background... :(

**Comment by ToniKorin** [comment]
According to the Apple Developer Technical Support CloudKit push notifications with user-visible keys such as alert, sound, or badge are NOT throttled. And Apple just published Web Service and JavaScript APIs for CloudKit.

**Comment by clarklight** [comment]
I am wonder does the Android now have the silent pass through now?

**Comment by clarklight** [comment]
Does it compile the the AppStore? As they are touchy with the background mode and their guidelines, wondering if the Background triggering an Ajax call will get accepted by the Appstore.

**Comment by ToniKorin** [comment]
You have no problems with the acceptance using the pure Ajax calls. In addition my app "Team Locator" is making a location update triggered by the notification and it got also accepted.

Unfortunately the reliability/throttling is the biggest issue here. I created a feature/enhancement request for Apple about the throttling. You should also do that, because extra requests will increase probability to get it done in some day.

https://bugreport.apple.com/

**Comment by clarklight** [comment]
@ToniKorin Thanks! I will do that. I am wondering about the reliability, if it is very unreliable? Or roughly do you have a percentage of the APNS to pass through and triggers an Ajax. Because if it is a problem, at the moment my only solution is to build a feedback loop back to the server to tell me that the APNS has been received, otherwise timeout it and resend.

**Comment by ToniKorin** [comment]
When your mobile has a weak network connection (-100dBm caused by e.g. indoors or car driving) it usually doesn't help to resend the notification. You can read more people's experience from there:
https://devforums.apple.com/message/1069621#1069621

**Comment by clarklight** [comment]
For anyone that are still looking for Android Silent pass through and have not figured it out. In GCMIntentService.java
Go to the part that says //Send a notification if there is a message.
in that If statment 
add in       PushPlugin.sendExtras(extras); 
That would add in the notification in the background(as the extras is the message string). So if you want to remove the notification to make it silent take out the createNotification part.

**Comment by ToniKorin** [comment]
I published a new version 1.1.1 of this plugin including BG support for Android into the plugins repo:
http://plugins.cordova.io/#/package/com.tonikorin.cordova.plugin.pushplugin
 It can trigger either the "sendEtxtras" call to your WebView running in BG (it requires e.g. katzer's Great Cordova Background Plug-in)  or start your native Android service. If you need to support older  Android versions than 4.0.4 please use  the version 1.1.0. Only difference is the permission requirement of GET_ACCOUNTS. The PushPlugin has two new configuration options in the register function related to the BG logic (false by default):

"receiveNotificationInBackground": true, => triggers sendEtxtras to your running WebView in BG
"startServiceAlwaysInBackground": true => start Android service even you have WebView running

And if you don't want a visible notification in BG, just leave the "message" property out and use your own property for the data (old functionality).

The Class Name of your native BG service is fetched from the notification property "service". The native Android service consumes less memory, less cpu and it is triggered even your app is not running.  If you like to implement a native Android Service of your own, It might help you to check my example of it from (also now in use in my Team Locator app):
http://plugins.cordova.io/#/package/com.tonikorin.cordova.plugin.locationprovider

BTW, so far GCM seems to be reliable and no delays in my tests, even I found some other experiences in several forums. :)

**Comment by clarklight** [comment]
I am wondering if silent pass through works if the App is closed, can it trigger ECB and a function if the app is totally closed off?

**Comment by markusv** [comment]
thanks for this. It works great! however I have trouble doing an jquery ajax request on response to a action on a push message. It seems the ajax request is stopped by the device. When the user starts the app (by clicking the icon on the homescreen) immediately the request finishes. Can one run a ajax request in response to a users clik on one action on a push message?

**Comment by clarklight** [comment]
@ToniKorin I checked out your version, wondering you added in the Silent(background) passthru for IOS? Because it isnt working on mine. I still have to click on the notification to trigger the ECB on IOS.

**Comment by ToniKorin** [comment]
Yes, it is included. I guess you just faced the throttling of iOS. If you
add some visible notification items like message or sound the iOS will more
likely wake your app.

clarklight <notifications@github.com> kirjoitti 27.7.2015 kello 9.13:

@ToniKorin <https://github.com/ToniKorin> I checked out your version,
wondering you added in the Silent(background) passthru for IOS? Because it
isnt working on mine. I still have to click on the notification to trigger
the ECB on IOS.

—
Reply to this email directly or view it on GitHub
<https://github.com/phonegap-build/PushPlugin/issues/288#issuecomment-125099167>
.

**Comment by clarklight** [comment]
I just tried adding in a custom sound in  but it doesnt seem to be triggering the background notification. And you mentioned before the throttling occurs mostly when device is not on Wifi right? Because i am plugged into power running on wifi next to the router. Not sure if there are other factors that might affect it. 

     if (e.sound) {
                var snd = new Media(e.sound);
                snd.play();
                 var my_media = new Media('/media/sound.mp3');
               my_media.play();  
           
            }

**Comment by clarklight** [comment]
Very Odd, my IOS version still isnt triggering, didnt trigger once. I even put an alert and console.log in the backgroundDone function, it isnt getting triggered(success or onError). 

PushNotification.prototype.backgroundDone = function(successCallback, errorCallback) {
    if (errorCallback == null) { errorCallback = function() {
        // Try to add in the Javascript function here!!!!
        alert("errorsync1");
          console.log("PushNotification.backgroundDone failure: success callback parameter must be a function");
         // iossync();
    }}

    if (typeof successCallback != "function") {
          //iossync();
          alert("success sync");
        console.log("PushNotification.backgroundDone failure: success callback parameter must be a function");
        return
    }

     cordova.exec(successCallback, errorCallback, "PushPlugin", "didCompleteBackgroundProcess", []);
}

**Comment by kevindonahue** [comment]
Hi clarklight,  If everything works successfully, neither or your alert messages will get called.  The second alert will only occur if you haven't defined a successCallback function (successCallback != "function").  Have you tried putting an alert in just before cordova.exec? Also, if you suspect that the push notifications aren't coming through, try and get access to the log files generated by the NSLog commands in the plugin.  But as ToniKorin stated, its probably due to iOS throttling.

**Comment by clarklight2** [comment]
Hi @kevindonahue I am not 100% sure where should i set the javascript function that i want it to trigger. within the     window.onNotificationAPN = function(e) { // the function goes here?}

But i am receiving the push notification. It buzz the device and play sound for the notification. But when i click the App icon to open the app, the function isnt triggered. I still have to click the notification to trigger it. 

Not sure what am i doing wrong.

**Comment by kevindonahue** [comment]
Been a while since I've worked on this, but I believe yes, you just call it once, after register:

```
function onNotificationAPNSuccess(e) {
  //after register was successful
  window.plugins.pushNotification.backgroundDone(backgroundCompleteSuccess, function (){ alert('failed'); });
}

function backgroundCompleteSuccess() {
  alert('success');
  //whatever you want to do now
}
```

**Comment by clarklight2** [comment]
Yer i am not able to get the background pass-through even calling it. I had been told that i would need a content-available key in the server side??

**Comment by kevindonahue** [comment]
Yes, your payload (push message) needs to have the content-available flag:

{ "aps" : { "alert" : "", "content-available" : 1 } , "extra_stuff" : "[your message stuff here]"  }

**Comment by clarklight2** [comment]
@kevindonahue Hey Kevin! Thanksss, yer it was the content-available tag..... now everything works perfectly...

**Comment by cmartin81** [comment]
Does anyone have a fork of pushPlugin which has the method backgroundDone implemeted?
